accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+trivial to see that the first switch is exhaustive because every matched value
+will be an instance of `Pip` and the first case covers every single instance of
+`Pip`. (The second case is unreachable and pointless, but harmless.) Likewise,
+the second switch is clearly exhaustive because every value is a `Face` and the
+second case matches all instances of `Face`.
 
-```
-Face? - Face  expands to:  Face|Null - Face = Null
-Jack? - Null  expands to:  Jack|Null - Null = Jack
-```
+In the algorithm, when it looks at a space, it first *expands* it. If the type
+of a space is a sealed type, enum type, or `bool` (which is basically an enum
+type), then it replaces the space with a set of more precise spaces that cover
+the original space. Then it processes those each independently. If the cases are
+all exhaustive over every one of the expanded spaces, then they are exhaustive
+over the original unexpanded space too.
 
-Likewise, if the left type is `FutureOr<T>` for some type `T` and the right type
-is a subtype of `Future` or `T`, then expanding expands the `FutureOr<T>` to
-`Future<T>|T`.
+To expand a space `s` with type `T`, we create a new set of spaces that share
+the same properties as `s` but with possibly different types and restrictions.
+The resulting spaces are:
 
-```
-FutureOr<int> - int     expands to:  Future<int>|int - int = Future<int>
-FutureOr<int> - Future  expands to:  Future<int>|int - Future = int
-```
+*   **The declaration `D` of `T` is a `sealed` type:**
 
-### Subtraction
+    1.  For each declaration `C` in the library of `D` that has `D` in an
+        `implements`, `extends`, or `with` clause:
 
-OK, that's enough preliminaries. Here's the algorithm. To subtract two object
-spaces `L - R`:
+        1.  If every type parameter in `C` forwards to a corresponding type
+            parameter in `D`, then `C` is a *trivial substitution* for `D`.
+            Let `S` be `C` instantiated with the type arguments of `T`.
 
-1.  If `L`'s type is sealed and `R`'s type is in its sealed subtype hierarchy,
-    then expand `L` to the union of subtypes and start the whole subtraction
-    process over. (That will then distribute the `- R` into all of the resulting
-    arms, process each subtype independently, and union the result.)
+            *For example:*
 
-2.  Else, if `R`'s type is not a subtype of `L`'s type (even after expanding)
-    then it can't meaningfully subtract anything. The result is just `L`. This
-    comes into play when when matching on unsealed subtypes:
+            ```dart
+            sealed class A<T1, T2> {}
+            class B<R1, R2> extends A<R1, R2> {}
+            ```
 
-    ```dart
-    class Animal {}
-    class Mammal extends Animal {}
+        2.  Else (not a trivial substitution) let `S` be an overapproximation
+            of `C`.
 
-    test(Animal a) {
-      switch (a) {
-        case Mammal m: ...
-      }
-    }
-    ```
+            **TODO:** Is the overapproximation part here correct?
```

Comment:
What *is* an "overapproximation"?

(Ah, found definition below. Add: `(see definition in "Overapproximation" section below)`)

The overapproximation is defined as a mapping from type to a type, but is probably really from a declaration to a type (types are instantiated, declarations are not), and `S` is a space, not a type. 
Should `S` be the space for that type?

Reply:
Reworded to be from a declaration to a type.

`S` is a type, not a space (may be a confusing choice of letter). We then create a space with type `S`.